### PR TITLE
On-device reductions

### PR DESCRIPTION
**Pull request by ngc92**
Moves loss calculation to backward, and ensures  we can do more on-device reductions and fewer host<->device transfers.
Also enables a micro-optimization, that validate does not calculate dlogits anymore.

### CODE REVIEW

File: llmc/cuda_common.h [hunk 1]
```diff
@@ -40,6 +41,10 @@ extern cudaDeviceProp deviceProp;
 // convenience macro for calculating grid/block dimensions for kernels
 #define CEIL_DIV(M, N) (((M) + (N)-1) / (N))
 
+// short-cuts for compile-time boolean values that can be used as function arguments
+constexpr std::bool_constant<true> True;
```
Comment (karpathy) [inline llmc/cuda_common.h:45]: I don't love bool_constant, it looks scary and caused an issue in an earlier PR, are we not able to avoid its use? Maybe just 0/1?

File: train_gpt2.cu [hunk 2]
```diff
@@ -686,45 +690,61 @@ void gpt2_forward(GPT2 *model, const int* inputs, const int* targets, size_t B,
     }
 
     matmul_forward_cublaslt(acts.output, acts.lnf, params.wte, NULL, B, T, C, Vp, main_stream);
+    cudaCheck(cudaDeviceSynchronize());
+}
 
-    // also forward the cross-entropy loss function if we have the targets
-    if (targets != NULL) {
+
+float gpt2_validate(GPT2 *model) {
```
Comment (karpathy) [inline train_gpt2.cu:697]: maybe a comment on what gpt2_validate does as a function? what is it validating exactly?

File: train_gpt2.cu [hunk 3]
```diff
@@ -883,12 +911,24 @@ void gpt2_backward_and_reduce(GPT2 *model, int* inputs, bool last_step) {
...
+        cudaCheck(cudaMemcpyAsync(&model->mean_loss, model->accumulated_mean_loss, sizeof(float), cudaMemcpyDeviceToHost, main_stream));
+
         floatX* const pointers[] = {grads.wte, grads.wpe, grads.lnfw, grads.lnfb};
         const size_t nelem[] = {Vp * C, T * C, C, C};
         multi_gpu_async_reduce_gradient(pointers, nelem, &multi_gpu_config, main_stream);
     }
 
     cudaCheck(cudaDeviceSynchronize());
+    if(last_step) {
+        model->mean_loss /= B*T*grad_accum_steps;
```
Comment (karpathy) [inline train_gpt2.cu:930]: this part also means that model->mean_loss is invalid/wrong if we try to read it out when !last_step

File: llmc/cuda_common.h (hunk 1 above)
Comment (ngc92) [reply llmc/cuda_common.h:45]: not really. The point of bool constant is that it encodes true/false in the _type_, so it can be used for function overloading/template argument deduction.
The alternative is to specify template arguments by hand, leading to
```cpp
kernel<<<block, grid>>><true>(args)
```
which is arguably much worse than bool_constant.
(or the other way round, not sure if <<<<>>>  or <> comes first).
The issue in the other PR was that we were not including the standard header where it is defined, so it wasn't really anything specific to bool_constant.

File: train_gpt2.cu (hunk 2 above)
Comment (ngc92) [reply train_gpt2.cu:697]: yes, the name isn't great. Maybe we should actually move the gpt2_forward call for validation inside here, so that the API allows for a future optimization in which the forward pass doesn't store all activations, and thus can run much larger batch size in validation.

File: train_gpt2.cu [hunk 4]
```diff
@@ -686,45 +690,61 @@ void gpt2_forward(GPT2 *model, const int* inputs, const int* targets, size_t B,
     }
 
     matmul_forward_cublaslt(acts.output, acts.lnf, params.wte, NULL, B, T, C, Vp, main_stream);
+    cudaCheck(cudaDeviceSynchronize());
```
Comment (karpathy) [inline train_gpt2.cu:693]: curious why

File: llmc/cuda_common.h (hunk 1 above)
Comment (karpathy) [reply llmc/cuda_common.h:45]: I guess my question is why do we need template arguments for a switch

File: train_gpt2.cu (hunk 4 above)
Comment (ngc92) [reply train_gpt2.cu:693]: it might not be necessary, but  I think for now, a programming model where every gpt2_... function guarantees that all its work is done by the time the function returns hopefully reduces the chances for errors.

File: llmc/cuda_common.h (hunk 1 above)
Comment (ngc92) [reply llmc/cuda_common.h:45]: by having it as a template argument, we make the compiler generate two specialized versions of the function that can be optimized specifically for the task at hand. The performance gains are _probably_ really minor/below the measurement noise.

File: train_gpt2.cu (hunk 3 above)
Comment (ngc92) [reply train_gpt2.cu:930]: yes, we can cheaply set it to -1 in the else branch to mark it as unavailable
